infrastructure/lib/constructs/api-function.ts

```diff
@@ -45,13 +41,22 @@ export abstract class ApiFunction extends cdk.Construct {
    */
   public readonly route: apigw.Method;
 
+  /**
+   * The CfnFunction logicalId, used for templating the api yaml.
+   */
+  public readonly logicalId: lambda.CfnFunction;
+
   protected constructor(scope: cdk.Construct, id: string, props: ApiFunctionProps) {
     super(scope, id);
     this.method = props.method;
     this.fn = new lambdaNodeJs.NodejsFunction(this, "Function", {
       entry: props.handlerPath,
       ...props.functionPropsOverride,
     });
-    this.route = props.resource.addMethod(props.method, new apigw.LambdaIntegration(this.fn));
+    this.fn.addPermission("AllowApiGatwewayInvoke", {
+      principal: new iam.ServicePrincipal("apigateway.amazonaws.com"),
+    });
+    this.logicalId = this.fn.node.defaultChild as lambda.CfnFunction;
```

Comment:
I am hesitant to call this `logicalId` because it isn't. I also very much don't want to make it publicly accessible since it can be derived from `this.fn`. Can we call it `const functionL1` (for Layer 1 construct) or `const cfnFunction` or `const rawCloudFormationFunction`? `Cfn` is often used for "CloudFormation" in the CDK and I see no reason to diverge from that pattern.

Reply:
While you're changing class attributes, we may want to change `this.fn`'s type to be `lambda.IFunction` instead of `lambda.Function`.

Reply:
```
lib/constructs/api-dynamodb-function.ts(30,13): error TS2339: Property 'addEnvironment' does not exist on type 'IFunction'.
lib/constructs/api-s3-function.ts(18,13): error TS2339: Property 'addEnvironment' does not exist on type 'IFunction'.
```
cdk isn't playing well with `lambda.IFunction`

Reply:
Good catch. There's a solution to this but it's not worth it. Continue using `lambda.Function`